packages/aggregator/src/__tests__/analytics/aggregateAnalytics/aggregations/getFinalValuePerPeriod.test.js

```diff
@@ -0,0 +1,128 @@
+        dataElement: 'element1',
+        organisationUnit: 'org2',
+        value: 2,
+      };
+
+      const startDateOrg1 = '20210101';
+      const startDateOrg2 = '20210303';
+      const aggregationPeriod = 'WEEK';
+      const testAnalytics = [
+        { ...defaultVariablesForOrg1, period: startDateOrg1 },
+        { ...defaultVariablesForOrg2, period: startDateOrg2 },
+      ];
+      const aggregationConfig = {
+        fillEmptyPeriodsWith: 'previous',
+      };
+
+      expect(
+        getFinalValuePerPeriod(testAnalytics, aggregationConfig, aggregationPeriod),
+      ).toStrictEqual([
+        ...getExpectedArray(startDateOrg1, aggregationPeriod, defaultVariablesForOrg1),
+        ...getExpectedArray(startDateOrg2, aggregationPeriod, defaultVariablesForOrg2),
+      ]);
+    });
+
+    it(`Filling preset value by 'fillEmptyPeriodsWith'`, () => {
+      const testAnalytics = [
+        { dataElement: 'element1', organisationUnit: 'org1', period: '20210101', value: 1 },
+        { dataElement: 'element1', organisationUnit: 'org1', period: '20210104', value: 2 },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '20210105', value: 1 },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '20210107', value: 3 },
+      ];
+      const aggregationPeriod = 'DAY';
+      const aggregationConfig = {
+        fillEmptyPeriodsWith: 'test',
+      };
+      expect(
+        getFinalValuePerPeriod(testAnalytics, aggregationConfig, aggregationPeriod),
+      ).toStrictEqual([
+        // org1
+        { dataElement: 'element1', organisationUnit: 'org1', period: '20210101', value: 1 },
+        { dataElement: 'element1', organisationUnit: 'org1', period: '20210102', value: 'test' },
+        { dataElement: 'element1', organisationUnit: 'org1', period: '20210103', value: 'test' },
+        { dataElement: 'element1', organisationUnit: 'org1', period: '20210104', value: 2 },
+        { dataElement: 'element1', organisationUnit: 'org1', period: '20210105', value: 'test' },
+        { dataElement: 'element1', organisationUnit: 'org1', period: '20210106', value: 'test' },
+        { dataElement: 'element1', organisationUnit: 'org1', period: '20210107', value: 'test' },
+        // org2
+        { dataElement: 'element1', organisationUnit: 'org2', period: '20210101', value: 'test' },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '20210102', value: 'test' },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '20210103', value: 'test' },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '20210104', value: 'test' },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '20210105', value: 1 },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '20210106', value: 'test' },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '20210107', value: 3 },
+      ]);
+    });
```

Comment:
@billli0 I don't think this is correct. I believe we should only by filling missing values within found values **per** org unit. eg:
```suggestion
    it(`Filling preset value by 'fillEmptyPeriodsWith'`, () => {
      const testAnalytics = [
        { dataElement: 'element1', organisationUnit: 'org1', period: '20210101', value: 1 },
        { dataElement: 'element1', organisationUnit: 'org1', period: '20210104', value: 2 },
        { dataElement: 'element1', organisationUnit: 'org2', period: '20210105', value: 1 },
        { dataElement: 'element1', organisationUnit: 'org2', period: '20210107', value: 3 },
      ];
      const aggregationPeriod = 'DAY';
      const aggregationConfig = {
        fillEmptyPeriodsWith: 'test',
      };
      expect(
        getFinalValuePerPeriod(testAnalytics, aggregationConfig, aggregationPeriod),
      ).toStrictEqual([
        // org1
        { dataElement: 'element1', organisationUnit: 'org1', period: '20210101', value: 1 },
        { dataElement: 'element1', organisationUnit: 'org1', period: '20210102', value: 'test' },
        { dataElement: 'element1', organisationUnit: 'org1', period: '20210103', value: 'test' },
        { dataElement: 'element1', organisationUnit: 'org1', period: '20210104', value: 2 },
        // org2
        { dataElement: 'element1', organisationUnit: 'org2', period: '20210105', value: 1 },
        { dataElement: 'element1', organisationUnit: 'org2', period: '20210106', value: 'test' },
        { dataElement: 'element1', organisationUnit: 'org2', period: '20210107', value: 3 },
      ]);
    });
```

Is there a reason you want the behaviour to be as you have written in this test?

Reply:
Hi @rohan-bes, I agree we should fill in missing value only for the gap, but when we get the period from [getContinuousPeriodsForAnalytics()](https://github.com/beyondessential/tupaia/blob/e6b5847088de0a5c677b586889cec2657e1ac794/packages/aggregator/src/analytics/aggregateAnalytics/aggregations/utils/getContinuousPeriodsForAnalytics.js#L16), we will get the earliest and the latest date from analytics, which doesn't count on the orgUnit and dataElement. 

If that's the right behaviour, do you suggest we should make it count on the orgUnit and dataElement? I guess it should be simple.

Reply:
Ahh I see your point. Yeah I'd say it'd be best then to recalculate the `periods` using `getContinuousPeriodsForAnalytics()` per orgUnit and dataElement.